### ISSUE TITLE
Creating RDF documents as one of the core functionalities

### ISSUE THREAD

**Issue by rajaram5** [OpenRefine/OpenRefine#4518]
Hello,
I am a software developer/data architect. I am involved in many research projects where we carry out data uplifting activities. Our activities including converting source data into `RDF`. We use `openrefine + rdf extension` to create RDF documents in our project. I wonder if there is a plan for the `openrefine team` to support creating RDF documents as one of the core functionalities of openrefine.

**Comment by andrawaag** [comment]
In the [past](https://moh-it.pure.elsevier.com/en/publications/the-fairification-of-data-and-the-potential-of-fair-resources-dem) I worked with @rajaram5 on making data FAIR using Open Refine with an extension. I agree with Rajaram that making this functionality core to OpenRefine would be a major enhancement.

**Comment by ostephens** [comment]
Would this be served by including the RDF extension as part of the OpenRefine distribution or is this a request for different functionality to the current extension?

**Comment by rajaram5** [comment]
To start with it would be good idea to include RDF extension as part of OpenRefine distribution with some UI fixes to the RDF skeleton of RDF extension

**Comment by ostephens** [comment]
I tend to agree with your view @wetneb - if we could deliver a good "extension management" that could be controlled by the user then the question of what was 'core' and what was not would become much less of an issue generally and that would be a better approach than just bringing all useful extensions under central maintenance

My only point against this is that adding the RDF extension to the distribution is something that would be pretty easy to do and although I completely accept your points about the maintenance and responsibility - this would raise the profile of the RDF extension, and if @stkenny was willing to maintain within the OR repo instead of separately then we'd basically be in the same situation as now with respect to maintenance - but with a raised profile for the RDF extension which could make it easier to get more developers involved

**Comment by wetneb** [comment]
Unfortunately, even if @stkenny agrees to maintain the extension as he is doing now, it does change quite a lot of things:

Say I want to introduce a breaking change in some API (for instance the change of JSON serialization from org.json to Jackson, to take a dramatic example). I cannot introduce it in OpenRefine without also updating the RDF extension in the same go, otherwise the CI build for this repository will fail. So either I do it (tada! I am now maintaining that extension too!), or I want to stick to my current commitments and not touch the RDF extension, what should I do? Leave the PR open and ask @stkenny to do the upgrade in the RDF extension in the same PR as soon as possible? That means putting a lot of pressure on him or a lot of frustration on me.

Many of our changes are time-sensitive: we need to fix an important vulnerability (say log4shell), we need to unblock the work of a contractor (say the Debian packaging), we need to release features from a funded project, and so on.

Similarly, if @stkenny wants to release a new version of the RDF extension, he can also be blocked by our agenda in the rest of the tool.

In the current situation, @stkenny can decide to update the RDF extension at his own pace instead.

**Comment by thadguidry** [comment]
Nothing more to add to @wetneb comments.  He's spot on with the direction that we should approach for both extension authors, our users, and our maintainers.

**Comment by andrawaag** [comment]
As much as I agree with the Idevops_ points discussed so far, there is another angle I would like to add to the discussion. From a user perspective, navigating the version control between Openrefine and its extensions can be complicated at times. Each extension will need to be updated which each new release of open refine. Then there is the recent development where Open Refine (and other packages) are packed in portals such as, for example, the [Jupyter Notebook](https://hub.paws.wmcloud.org/) server of the Wikimedia Foundation [1]. This is huge since it allows easier access. Unless of course enabling  and maintaining plugins is part of the UX and does not require admin access to for example PAWS. To my knowledge that isn't currently the case. 



[1] https://hub.paws.wmcloud.org/

**Comment by thadguidry** [comment]
@andrawaag 
If we developed a feature where an Extensions manager could install extensions over the internet if users were given a dialog prompt to allow installing over the internet or not...
1. Do you foresee issues where the UX of the Extensions manager showed which extensions were compatible (regular text - checkbox is available to install them ) and which were not (greyed text - no checkbox to select and install them)?
2. On the manual side of things, Do you foresee issues where a local folder path for extensions can be specified and then OpenRefine checks that path upon startup as it does now and shows the Extensions manager and warnings about incompatible extensions that could not be loaded?

**Comment by stkenny** [comment]
Unfortunately I think having a vibrant ecosystem of extensions, whilst also keeping the ability to rapidly introduce breaking API changes may well be mutually exclusive. At a minimum the extension developers will need some notice of what needs to be changed.

Other than the Jackson change the RDF extension hasn't really changed much since it was forked from Fadi Maali's repository. It has really just been trying to keep it working, so it could do with some new features/design. Having it in the OpenRefine organisation would likely be beneficial as it would at least give a single authoritative source for people. The number of forks and outdated links has caused confusion in the past. See for example https://github.com/stkenny/grefine-rdf-extension/issues/48

**Comment by wetneb** [comment]
> Unfortunately I think having a vibrant ecosystem of extensions, whilst also keeping the ability to rapidly introduce breaking API changes may well be mutually exclusive. At a minimum the extension developers will need some notice of what needs to be changed.

I have tried to provide [migration guidelines](https://github.com/OpenRefine/OpenRefine/wiki/Migration-guide-for-extension-and-fork-maintainers) in the past and am very keen to know how it can be done better.
It would really help me to have concrete feedback on this! 

Also, making it easier to install extensions and detect their incompatibilities with OpenRefine is likely to require some breaking changes on its own (such as: migrating out of Butterfly to something else), so that is a bit of a dilemma for me. Here too, input welcome!

> Having it in the OpenRefine organisation would likely be beneficial as it would at least give a single authoritative source for people.

Okay, given the other comments in this thread I think we have a consensus for that, so let us do it :) I will create the repo and invite you.

**Comment by AtesComp** [comment]
I created RDF Transform to combine the RDF extension with features of the OntoRefine version and extend with more robust processing.  Review to see if this suits many of the issues reported here.

See: [RDF Transform](https://github.com/AtesComp/rdf-transform)
